sealevel-ffi/sealevel.h

```diff
@@ -0,0 +1,267 @@
+
+/**
+ * The invoke context holds the state of a single transaction execution.
+ * It tracks the execution progress (instruction being executed),
+ * interfaces with account data,
+ * and specifies the on-chain execution rules (precompiles, syscalls, sysvars).
+ */
+typedef struct sealevel_invoke_context sealevel_invoke_context;
+
+typedef struct sealevel_vm sealevel_vm;
+
+/**
+ * The map of syscalls provided by the virtual machine.
+ */
+typedef void *sealevel_syscall_registry;
+
+typedef struct {
+  void *data_addr;
+  size_t data_size;
+  uint64_t vm_addr;
+  uint64_t vm_gap_size;
+  bool is_writable;
+} sealevel_region;
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+
+/**
+ * Creates a new Sealevel machine config.
+ *
+ * # Safety
+ * Call `sealevel_config_free` on the return value after you are done using it.
+ * Failure to do so results in a memory leak.
+ */
+sealevel_config *sealevel_config_new(void);
+
+/**
+ * Sets a config option given the config key and exactly one value arg.
+ *
+ * # Safety
+ * Avoid the following undefined behavior:
+ * - Passing the wrong argument type as the config value (each key documents the expected value).
+ */
+void sealevel_config_setopt(sealevel_config *config, sealevel_config_opt key, size_t value);
+
+/**
+ * Releases resources associated with a Sealevel machine config.
+ *
+ * # Safety
+ * Avoid the following undefined behavior:
+ * - Calling this function given a string that's _not_ the return value of `sealevel_config_new`.
+ * - Calling this function more than once on the same object (double free).
+ * - Using the config object after calling this function (use-after-free).
+ */
+void sealevel_config_free(sealevel_config *config);
+
+/**
+ * Creates a new, empty syscall registry.
+ *
+ * # Safety
+ * Call `sealevel_syscall_registry_free` on the return value after you are done using it.
+ * Failure to do so results in a memory leak.
+ */
+sealevel_syscall_registry sealevel_syscall_registry_new(void);
+
+/**
+ * Frees a syscall registry.
+ *
+ * # Safety
+ * Avoid the following undefined behavior:
+ * - Calling this function twice on the same object.
+ * - Using the syscall registry after calling this function.
+ */
+void sealevel_syscall_registry_free(sealevel_syscall_registry registry);
+
+/**
+ * Drops an invoke context and all programs created with it. Noop given a null pointer.
+ *
+ * # Safety
+ * Avoid the following undefined behavior:
+ * - Calling this function twice on the same object.
+ * - Using the invoke context after calling this function.
+ */
+void sealevel_invoke_context_free(sealevel_invoke_context *this_);
+
+/**
+ * Returns the error code of this thread's last seen error.
+ */
+int sealevel_errno(void);
+
+/**
+ * Returns a UTF-8 string of this thread's last seen error,
+ * or NULL if `sealevel_errno() == SEALEVEL_OK`.
+ *
+ * # Safety
+ * Call `sealevel_strerror_free` on the return value after you are done using it.
+ * Failure to do so results in a memory leak.
```

Comment:
This sucks. Error strings should be written to a thread-local bss buffer instead.

Reply:
Fixed.